Make QMK actually work as per the official documentation

Ekleog opened the pull request:
### nixpkgs-check report

**version:** `nixpkgs-check v0.1.0` on NixOS 21.05, sandbox = "true"
**packages declared changed:** {"python3Packages.pyrsistent", "qmk", "qmk-udev-rules", "python3Packages.pyusb"}
**manual tests declared performed:**
 * ✔ built on NixOS
 * 😢 not built on MacOS
 * 😢 not built on Other Linux distributions
 * manually compile and flash a moonlander keyboard as per the official documentation

**complies with contributing.md:** ✔ yes
**package python3Packages.pyrsistent:** ✔ continued building
**package qmk:** ✔ continued building
**package qmk-udev-rules:** 💚 started building again
**package python3Packages.pyusb:** ✔ continued building
**closure size for python3Packages.pyrsistent:** ✔ increased by 2.6 KB, from 98.8 MB to 98.8 MB
**tests of python3Packages.pyrsistent:** 😢 there are no tests

**closure size for qmk:** 😢 increased by 2.4 GB, from 181.5 MB to 2.6 GB
**tests of qmk:** 😢 there are no tests
**binaries of qmk:**
  * *updated binaries:*
    * ✔ qmk continued running successfully


**tests of qmk-udev-rules:** 😢 there are no tests

**closure size for python3Packages.pyusb:** ✔ increased by 2.6 KB, from 150.5 MB to 150.5 MB
**tests of python3Packages.pyusb:** 😢 there are no tests

Ekleog committed:
qmk: add all dependencies required for normal operation

r-rmcgibbo commented:
Result of `nixpkgs-review pr 130847` at 1f60c33a run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>deeptools</li>
    <li>python38Packages.datashader</li>
    <li>python38Packages.spacy</li>
    <li>python38Packages.spacy-transformers</li>
    <li>python38Packages.textacy</li>
    <li>spyder</li>
  </ul>
</details>
<details>
  <summary>31 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/1d4cc7f0f307e718b357776d54e16699">aws-sam-cli</a></li>
    <li><a href="https://gist.github.com/6cae323082ee0ccec2392c2f02f5e462">gnuk</a></li>
    <li><a href="https://gist.github.com/02b49cb537f4c6ceb20cc9b96e3f42fc">python38Packages.adext</a></li>
    <li><a href="https://gist.github.com/5346897cd186ab4f52a6a48c3afb100b">python38Packages.alarmdecoder</a></li>
    <li><a href="https://gist.github.com/0459978e982e54b5f7a0d25fe32435d8">python38Packages.flower</a></li>
    <li><a href="https://gist.github.com/789b70341c7d497f9505996db61d0a28">python38Packages.ipydatawidgets</a></li>
    <li><a href="https://gist.github.com/e160c88bad7a10b931e1d1c3a103cf04">python38Packages.kmapper</a></li>
    <li><a href="https://gist.github.com/d21f74c0f1e188e5484074b7c40789b1">python38Packages.pyftdi</a></li>
    <li><a href="https://gist.github.com/3f38d85dba45f8a8f2d778abf32dc13b">python38Packages.pyi2cflash</a></li>
    <li><a href="https://gist.github.com/a7db5b8e1a07defd924e0bfe1b404742">python38Packages.pymatgen-lammps</a></li>
    <li><a href="https://gist.github.com/847dabe676ee02d49b3b81e545f15ce4">python38Packages.pyspiflash</a></li>
    <li><a href="https://gist.github.com/6a7921024befc2de45bc8c4ddbf3ef45">python38Packages.robotframework-selenium2library</a></li>
    <li><a href="https://gist.github.com/bae599e8cbab61d58f843814b889d0a0">python38Packages.robotframework-seleniumlibrary</a></li>
    <li><a href="https://gist.github.com/1dee260b5355e5e987458596efce36e2">python38Packages.robotframework-tools</a></li>
    <li><a href="https://gist.github.com/6cb9e73c7016fb1c651c3f42debdda77">python38Packages.scikit-tda</a></li>
    <li><a href="https://gist.github.com/dfd3729c6ffc5017c369b6ef150b7d80">python38Packages.sopel</a></li>
    <li><a href="https://gist.github.com/148c8b64c9b3b0997aff80bbf95b15cf">python39Packages.adext</a></li>
    <li><a href="https://gist.github.com/d989cca83cf4e819fdf0da6dc28be6bb">python39Packages.alarmdecoder</a></li>
    <li><a href="https://gist.github.com/c960cf80eb174f1229d60e3bcd01c15d">python39Packages.flower</a></li>
    <li><a href="https://gist.github.com/d4c90822f50831c5265412c61278f857">python39Packages.ipydatawidgets</a></li>
    <li><a href="https://gist.github.com/f2088796a79531b27d3289a307129f8b">python39Packages.jupyterlab-git</a></li>
    <li><a href="https://gist.github.com/3e25b79d173d777423767c1442d798c5">python39Packages.kmapper</a></li>
    <li><a href="https://gist.github.com/4d707dae89e6bcfd616610bf1e129441">python39Packages.nbdime</a></li>
    <li><a href="https://gist.github.com/ef381467fd1eb52c703677bc3e4de82d">python39Packages.pyftdi</a></li>
    <li><a href="https://gist.github.com/8c982b6cc2f8657111ad75e70a2e3f74">python39Packages.pyi2cflash</a></li>
    <li><a href="https://gist.github.com/7ec6ec8f54210017e88b91abe29af13f">python39Packages.pymatgen-lammps</a></li>
    <li><a href="https://gist.github.com/692505cfb8738377d6f2247c36e9f06e">python39Packages.pyspiflash</a></li>
    <li><a href="https://gist.github.com/9940cf8e2105f0e44122eb03fff751ce">python39Packages.robotframework-selenium2library</a></li>
    <li><a href="https://gist.github.com/9628cb7deff0085379960f8405a8acaa">python39Packages.robotframework-seleniumlibrary</a></li>
    <li><a href="https://gist.github.com/0eb1c35a3d91e96c191af1cb1969a863">python39Packages.robotframework-tools</a></li>
    <li><a href="https://gist.github.com/443223d71774e2d90f14d6b26fd05d47">python39Packages.sopel</a></li>
  </ul>
</details>
<details>
  <summary>99 packages skipped due to time constraints:</summary>
  <ul>
    <li>ansible (ansible_2_11 ,python39Packages.ansible-core)</li>
    <li>ansible-lint (python39Packages.ansible-lint)</li>
    <li>ansible_2_10 (python39Packages.ansible-base)</li>
    <li>apache-airflow</li>
    <li>awscli2</li>
    <li>cloud-init</li>
    <li>cq-editor</li>
    <li>eliot-tree</li>
    <li>expliot</li>
    <li>flexget</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>203 packages built successfully:</summary>
  <ul>
    <li>anki</li>
    <li>ankisyncd</li>
    <li>antfs-cli</li>
    <li>arion</li>
    <li>avell-unofficial-control-center</li>
    <li>cloud-custodian</li>
    <li>cyclonedx-python</li>
    <li>docker-compose</li>
    <li>dt-schema</li>
    <li>g203-led</li>
    <li>gen-oath-safe</li>
    <li>krakenx</li>
    <li>liquidctl (python39Packages.liquidctl)</li>
    <li>matrix-commander</li>
    <li>nbstripout</li>
    <li>opsdroid</li>
    <li>pantalaimon</li>
    <li>pypi2nix</li>
    <li>python38Packages.BlinkStick</li>
    <li>python38Packages.Pweave</li>
    <li>python38Packages.apispec</li>
    <li>python38Packages.aws-sam-translator</li>
    <li>python38Packages.castepxbin</li>
    <li>python38Packages.celery</li>
    <li>python38Packages.cfn-lint</li>
    <li>python38Packages.chart-studio</li>
    <li>python38Packages.cve-bin-tool</li>
    <li>python38Packages.dash</li>
    <li>python38Packages.django-raster</li>
    <li>python38Packages.djmail</li>
    <li>python38Packages.facedancer</li>
    <li>python38Packages.flask-appbuilder</li>
    <li>python38Packages.flask-restplus</li>
    <li>python38Packages.geoip2</li>
    <li>python38Packages.greatfet</li>
    <li>python38Packages.holoviews</li>
    <li>python38Packages.hvplot</li>
    <li>python38Packages.ipympl</li>
    <li>python38Packages.ipyvue</li>
    <li>python38Packages.ipyvuetify</li>
    <li>python38Packages.ipywidgets</li>
    <li>python38Packages.jsonmerge</li>
    <li>python38Packages.jsonschema</li>
    <li>python38Packages.jupyter</li>
    <li>python38Packages.jupyter-sphinx</li>
    <li>python38Packages.jupyter-telemetry</li>
    <li>python38Packages.jupyter_server</li>
    <li>python38Packages.jupyterlab</li>
    <li>python38Packages.jupyterlab_launcher</li>
    <li>python38Packages.jupyterlab_server</li>
    <li>python38Packages.jupytext</li>
    <li>python38Packages.lightparam</li>
    <li>python38Packages.liquidctl</li>
    <li>python38Packages.matrix-nio</li>
    <li>python38Packages.mocket</li>
    <li>python38Packages.modeled</li>
    <li>python38Packages.moretools</li>
    <li>python38Packages.moto</li>
    <li>python38Packages.nbclassic</li>
    <li>python38Packages.nbconflux</li>
    <li>python38Packages.nbconvert</li>
    <li>python38Packages.nbformat</li>
    <li>python38Packages.nbmerge</li>
    <li>python38Packages.nbsmoke</li>
    <li>python38Packages.nbsphinx</li>
    <li>python38Packages.nbval</li>
    <li>python38Packages.notebook</li>
    <li>python38Packages.notedown</li>
    <li>python38Packages.openant</li>
    <li>python38Packages.openapi-core</li>
    <li>python38Packages.openapi-schema-validator</li>
    <li>python38Packages.openapi-spec-validator</li>
    <li>python38Packages.phik</li>
    <li>python38Packages.plotly</li>
    <li>python38Packages.pook</li>
    <li>python38Packages.prance</li>
    <li>python38Packages.pygreat</li>
    <li>python38Packages.pymatgen</li>
    <li>python38Packages.pyrsistent</li>
    <li>python38Packages.pyusb</li>
    <li>python38Packages.pyvisa-py</li>
    <li>python38Packages.rachiopy</li>
    <li>python38Packages.rfcat</li>
    <li>python38Packages.robomachine</li>
    <li>python38Packages.robotframework</li>
    <li>python38Packages.robotframework-databaselibrary</li>
    <li>python38Packages.robotframework-requests</li>
    <li>python38Packages.robotframework-sshlibrary</li>
    <li>python38Packages.robotstatuschecker</li>
    <li>python38Packages.robotsuite</li>
    <li>python38Packages.seabreeze</li>
    <li>python38Packages.solo-python</li>
    <li>python38Packages.sphinxcontrib-openapi</li>
    <li>python38Packages.swagger-spec-validator</li>
    <li>python38Packages.ttp</li>
    <li>python38Packages.usbtmc</li>
    <li>python38Packages.vega</li>
    <li>python38Packages.warlock</li>
    <li>python38Packages.webthing</li>
    <li>python38Packages.widgetsnbextension</li>
    <li>python38Packages.zetup</li>
    <li>python38Packages.zigpy-zigate</li>
    <li>python39Packages.BlinkStick</li>
    <li>python39Packages.Pweave</li>
    <li>python39Packages.altair</li>
    <li>python39Packages.apispec</li>
    <li>python39Packages.aws-sam-translator</li>
    <li>python39Packages.castepxbin</li>
    <li>python39Packages.celery</li>
    <li>python39Packages.cfn-lint</li>
    <li>python39Packages.chart-studio</li>
    <li>python39Packages.cve-bin-tool</li>
    <li>python39Packages.dash</li>
    <li>python39Packages.django-raster</li>
    <li>python39Packages.djmail</li>
    <li>python39Packages.facedancer</li>
    <li>python39Packages.flask-appbuilder</li>
    <li>python39Packages.flask-restplus</li>
    <li>python39Packages.flask-restx</li>
    <li>python39Packages.geoip2</li>
    <li>python39Packages.greatfet</li>
    <li>python39Packages.holoviews</li>
    <li>python39Packages.hvplot</li>
    <li>python39Packages.ipympl</li>
    <li>python39Packages.ipyvue</li>
    <li>python39Packages.ipyvuetify</li>
    <li>python39Packages.ipywidgets</li>
    <li>python39Packages.jsonmerge</li>
    <li>python39Packages.jsonschema</li>
    <li>python39Packages.jupyter</li>
    <li>python39Packages.jupyter-sphinx</li>
    <li>python39Packages.jupyter-telemetry</li>
    <li>python39Packages.jupyter_server</li>
    <li>python39Packages.jupyterlab</li>
    <li>python39Packages.jupyterlab_launcher</li>
    <li>python39Packages.jupyterlab_server</li>
    <li>python39Packages.jupytext</li>
    <li>python39Packages.lightparam</li>
    <li>python39Packages.matrix-nio</li>
    <li>python39Packages.mocket</li>
    <li>python39Packages.modeled</li>
    <li>python39Packages.moretools</li>
    <li>python39Packages.moto</li>
    <li>python39Packages.nbclassic</li>
    <li>python39Packages.nbconflux</li>
    <li>python39Packages.nbconvert</li>
    <li>python39Packages.nbformat</li>
    <li>python39Packages.nbmerge</li>
    <li>python39Packages.nbsmoke</li>
    <li>python39Packages.nbsphinx</li>
    <li>python39Packages.nbval</li>
    <li>python39Packages.notebook</li>
    <li>python39Packages.notedown</li>
    <li>python39Packages.openant</li>
    <li>python39Packages.openapi-core</li>
    <li>python39Packages.openapi-schema-validator</li>
    <li>python39Packages.openapi-spec-validator</li>
    <li>python39Packages.plotly</li>
    <li>python39Packages.pook</li>
    <li>python39Packages.prance</li>
    <li>python39Packages.pygreat</li>
    <li>python39Packages.pymatgen</li>
    <li>python39Packages.pyrsistent</li>
    <li>python39Packages.pyusb</li>
    <li>python39Packages.pyvisa-py</li>
    <li>python39Packages.rachiopy</li>
    <li>python39Packages.rfcat</li>
    <li>python39Packages.robomachine</li>
    <li>python39Packages.robotframework</li>
    <li>python39Packages.robotframework-databaselibrary</li>
    <li>python39Packages.robotframework-requests</li>
    <li>python39Packages.robotframework-sshlibrary</li>
    <li>python39Packages.robotstatuschecker</li>
    <li>python39Packages.robotsuite</li>
    <li>python39Packages.seabreeze</li>
    <li>python39Packages.solo-python</li>
    <li>python39Packages.sphinxcontrib-openapi</li>
    <li>python39Packages.swagger-spec-validator</li>
    <li>ttp (python39Packages.ttp)</li>
    <li>python39Packages.usbtmc</li>
    <li>python39Packages.vega</li>
    <li>python39Packages.warlock</li>
    <li>python39Packages.webthing</li>
    <li>python39Packages.widgetsnbextension</li>
    <li>python39Packages.zetup</li>
    <li>python39Packages.zigpy-zigate</li>
    <li>qmk-udev-rules</li>
    <li>snakemake</li>
    <li>sourcehut.buildsrht</li>
    <li>sourcehut.coresrht</li>
    <li>sourcehut.dispatchsrht</li>
    <li>sourcehut.gitsrht</li>
    <li>sourcehut.hgsrht</li>
    <li>sourcehut.hubsrht</li>
    <li>sourcehut.listssrht</li>
    <li>sourcehut.mansrht</li>
    <li>sourcehut.metasrht</li>
    <li>sourcehut.pastesrht</li>
    <li>sourcehut.todosrht</li>
    <li>streamlit</li>
    <li>tinyprog</li>
    <li>weechatScripts.weechat-matrix</li>
    <li>yubikey-manager</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/tools/misc/qmk/default.nix:3:3:
```
  |
3 | , fetchpatch
  |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

SuperSandro2000 commented on pkgs/os-specific/linux/qmk-udev-rules/default.nix:
```diff
@@ -0,0 +1,36 @@
+{ lib, stdenv, fetchFromGitHub }:
```
Why is this not an extra output in the qmk package?

Ekleog replied:
Because it's not the way it's being done for other `-udev-rules` I could find on the packages list, and also because the build is completely different (not the same repository, not the same version, etc.)

SuperSandro2000 commented on pkgs/tools/misc/qmk/default.nix:
```diff
@@ -27,13 +40,23 @@ buildPythonApplication rec {
     argcomplete
     colorama
-    dotty-dict
+    #dotty-dict # See https://pypi.org/project/qmk-dotty-dict/
```
```suggestion
```

Ekleog replied:
I'd rather keep this commented so that on the next update when qmk-dotty-dict becomes dotty-dict again, it's easier to make things work

SuperSandro2000 commented on pkgs/tools/misc/qmk/default.nix:
```diff
@@ -27,13 +40,23 @@ buildPythonApplication rec {
     pygments
+    pyrsistent
     pyusb
+  ] ++ [ # Binaries need to be in the path so this is in propagatedBuildInputs
```
Please wrap instead.

Ekleog replied:
There's already a wrapper in `buildPythonApplication` that adds things to the `PATH`, it doesn't make sense to me to wrap the same binary twice

SuperSandro2000 commented on pkgs/os-specific/linux/qmk-udev-rules/default.nix:
```diff
@@ -0,0 +1,36 @@
+    sha256 = "08d2ri9g6lky3ixd5h0scm95hgc52lcjr3vcnmpdxn780q9ygmkm";
+  };
+
+  buildPhase = ''
+    runHook preBuild
+    runHook postBuild
+  '';
+
```
```suggestion
  dontBuild = true;
```

Ekleog replied:
Great idea, thank you!

arcnmx commented:
This stands out as being an odd and fragile change, why would you not just use the existing nixpkgs cross-compile infrastructure by building `pkgsCross.avr.qmk` and so on instead?

Ekleog commented:
@arcnmx So this PR is just fixing the qmk tool as already existed, that supported only a small part of the features promised by the documentation.

The qmk ecosystem is split in two parts: the `qmk` tool, and the `qmk_firmware` repository.

The qmk tool is here to be able to fork, compile and flash the qmk_firmware repository. The qmk_firmware repository itself cannot reasonably be made into a nixpkgs package itself, because it is designed to be heavily modified by the user before being compiled.

So overall the qmk tool is similar to both cmake + openocd, with the detail that it directly depends on a known subset of gcc versions. Before this change only the openocd part worked, with this change the cmake part of it also starts working.

As you say, it would also be possible to have only the flashing part of it in nixpkgs and force the user to compile qmk_firmware themselves and then use qmk to flash it, which is approximately how things worked before this change. But then:
- there is basically no documentation about how to actually do the compiling (especially as it would also involve the cross part of nixpkgs that the end-user would have to learn how to use)
- we must still keep the qmk binary because it's the only official way to flash these keyboards, so users would end up with a binary for which some of the commands work and some of the commands don't (in particular the "setup" command doesn't work prior to this change)
- the user must compile the versions of gcc themselves, which is quite long even though it could have been pre-compiled by hydra as it's for a well-known very restricted subset of the architectures (qmk to the best of my knowledge only requires avr and arm, though I could test only the arm part of it as my keyboard is arm)

In a way it is true that it's close-ish to rustup, but ISTR rustup recently has been working in nixpkgs, and it requires quite a bit more understanding of nixpkgs to make it work manually without help than rustup in its basic usage pattern does, plus it's code designed to be edited while rustup is only made to download tarballs from mozilla and use them.

Hope this makes the reasoning for this clearer!

Now that the issues mentioned in this PR are fixed, I'm planning on rebasing to squash the new commits into the relevant old commits and merging in the next few days unless objections arise.

jtojnar reviewed (approved):
- `qmk doctor` is green.
- Verified that I can compile and flash my keymap into ARM-based Teensy 4.1 (with `teensy-loader-cli` on `PATH`).
- Changes look okay:
  - [pyusb 1.2.0](https://github.com/pyusb/pyusb/releases/tag/v1.2.0), [1.2.1](https://github.com/pyusb/pyusb/releases/tag/v1.2.1)
  - [pyrsistent 0.18.0](https://github.com/tobgu/pyrsistent/blob/v0.18.0/CHANGES.txt#L3-L15)
    - contains a [BC break](https://github.com/tobgu/pyrsistent/pull/209) but only has two other reverse dependencies:
      - [`jsonschema`](https://github.com/Julian/jsonschema) does not seem to use freeze/thaw
      - [`eliot`](https://github.com/itamarst/eliot) uses thaw but its tests still pass

sigprof commented:
Tested the package, and apparently compiling the firmware for an ARM-based keyboard (`planck/rev6`) works:
```
❯ qmk compile -kb planck/rev6 -km default
Ψ Compiling keymap with make -j 1 planck/rev6:default


QMK Firmware 0.13.34
Making planck/rev6 with keymap default

arm-none-eabi-gcc (GNU Arm Embedded Toolchain 10-2020-q4-major) 10.2.1 20201103 (release)
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Compiling: quantum/audio/muse.c                                                                     [OK]
...
Compiling: tmk_core/common/chibios/syscall-fallbacks.c                                              [OK]
Linking: .build/planck_rev6_default.elf                                                             [OK]
Creating binary load file for flashing: .build/planck_rev6_default.bin                              [OK]
Creating load file for flashing: .build/planck_rev6_default.hex                                     [OK]

Size after:
   text	   data	    bss	    dec	    hex	filename
      0	  52404	      0	  52404	   ccb4	.build/planck_rev6_default.hex

Copying planck_rev6_default.bin to qmk_firmware folder                                              [OK]
(Firmware size check does not yet support cortex-m4 microprocessors; skipping.)
```
However, compiling the firmware for an AVR-based keyboard (e.g., `planck/rev5`) fails:
```
❯ qmk compile -kb planck/rev5 -km default
Ψ Compiling keymap with make -j 1 planck/rev5:default


QMK Firmware 0.13.34
Making planck/rev5 with keymap default

avr-gcc (GCC) 8.4.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Compiling: quantum/audio/muse.c                                                                     [OK]
...
Compiling: tmk_core/protocol/lufa/usb_util.c                                                        [OK]
Linking: .build/planck_rev5_default.elf                                                             [OK]
Creating load file for flashing: .build/planck_rev5_default.hex                                     [ERRORS]
 |
 | sh: avr-objcopy: command not found
 |
make[1]: *** [tmk_core/rules.mk:296: .build/planck_rev5_default.hex] Error 1
Make finished with errors
make: *** [Makefile:539: planck/rev5:default] Error 1
```
Apparently the problem here is that `avr-objcopy` is not directly provided by the `pkgsCross.avr.buildPackages.binutils` package, which is actually a wrapper — that executable is contained in the real `binutils` package, which is listed in `nix-support/propagated-user-env-packages` inside that `binutils-wrapper`. This is not a problem when `pkgsCross.avr.buildPackages.binutils` is used inside `mkShell` (therefore `shell.nix` inside `qmk_firmware` works), but apparently `buildPythonApplication` does not pick up extra PATH elements from `nix-support/propagated-user-env-packages` files of packages listed in `propagatedBuildInputs`, therefore the resulting PATH is incomplete and does not contain all required tools.

I don't know what would be a proper generic way to fix this problem, but I have been able to fix this particular instance by adding `pkgsCross.avr.buildPackages.binutils.bintools` to `propagatedBuildInputs` (it must be after `pkgsCross.avr.buildPackages.binutils` to make the wrapper work properly). The other thing that is added to `nix-support/propagated-user-env-packages` in that package is `.libc` (= `pkgsCross.avr.libcCross`), but it is already in the list.

sigprof commented on pkgs/tools/misc/qmk/default.nix:
```diff
@@ -27,13 +39,23 @@ buildPythonApplication rec {
+    dfu-programmer
+    dfu-util
+    gcc-arm-embedded
+    pkgsCross.avr.buildPackages.binutils
```
```suggestion
    pkgsCross.avr.buildPackages.binutils
    pkgsCross.avr.buildPackages.binutils.bintools
```
See my comment: https://github.com/NixOS/nixpkgs/pull/130847#issuecomment-899090604 — AVR firmwares don't compile without this.

sigprof commented on pkgs/tools/misc/qmk/default.nix:
```diff
@@ -27,13 +39,23 @@ buildPythonApplication rec {
+  ] ++ [ # Binaries need to be in the path so this is in propagatedBuildInputs
+    avrdude
+    dfu-programmer
+    dfu-util
```
```suggestion
    dfu-util
    teensy-loader-cli
```
Please also add `teensy-loader-cli`, so that Teensy boards can be flashed (the tool is tiny and does not have any dependencies except libusb, which is already in the dependency tree).

sigprof reviewed (changes requested):
Looks great — I was able to compile and flash a firmware for my AVR-based board (`id80/ansi`) with this, but needed to fix some problems (noted in suggestions). The `qmk console` command also works (both for the console messages and for the bootloader connection notifications).

BTW, `handwired/onekey/*` “keyboards” should cover a representative set of supported MCUs and bootloaders, so you can use a command like this for testing whether the build environment works (you don't actually need the hardware for that):
```sh
for k in $(cd "$(qmk env QMK_HOME)"/keyboards && ls handwired/onekey/*/rules.mk | sed 's,/rules\.mk$,,'); do qmk compile -kb "$k" -km default; done
```

Ideally you should also run `qmk flash` for the same set of keyboards and make sure that you get some kind of a “waiting for bootloader” message (in many cases, except mostly the `caterina` bootloader, this actually runs the corresponding flashing tool, so it would fail if the tool is not present); doing this even without the actual hardware could be useful to confirm that all tools are in place.

jtojnar commented on pkgs/tools/misc/qmk/default.nix:
```diff
@@ -27,13 +39,23 @@ buildPythonApplication rec {
     jsonschema
     milc
     pygments
+    pyrsistent
```
https://github.com/NixOS/nixpkgs/pull/136976 updated qmk in Nixpkgs and it looks like pyrsistent is no longer needed: https://github.com/qmk/qmk_cli/commit/8e1de7491882cccaa7d7c66fbad0c0c0cd059474

happysalada commented:
I didn't know this PR existed before merging the qmk cli update.
Do you think you would be able to rebase ?
I have a darwin machine and I am happy to test on darwin.

happysalada commented:
Regarding closure size, I was wondering if we couldn't implement a `features` options where people could choose which dependencies they would want (depending on which board they had. Of course if people wanted to flash multiple boards, they could choose all the features).
Note that this is just a suggestion and does not need to be implemented in that PR or at all for that matter.

Ekleog commented:
… ok I had completely forgotten about this PR, sorry about it! So I've rebased to latest master, removed all the miscellaneous package bumps that already made it in, and removed pyrsistent which is apparently no longer needed indeed. Given the feedback on the PR I'll land as is, and we can then iterate to eg. reduce the closure size, though I'd maybe lean more towards all-features-by-default unless someone gets motivated enough to add nixos-specific error messages to explain how to make things work again :)

Ekleog committed:
qmk: also add teensy flash tool

Co-authored-by: Sergey Vlasov <sigprof@gmail.com>

Ekleog committed:
qmk: make sure AVR-based keyboards can build too

Co-authored-by: Sergey Vlasov <sigprof@gmail.com>
